### PR TITLE
Add more docs and examples for ListArray and OffsetsBuffer

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
N/A

# Rationale for this change
 
I personally find `ListArray` confusing both for myself and to explain to others. Recently I made some pictures while reviewing and discussing with @izveigor  in DataFusion https://github.com/apache/arrow-datafusion/pull/6936#discussion_r1279718073 and want to add them to the docs so I can keep referring to them 

I also think it would be nice to be able to use ListArray without having to refer to the Arrow spec

# What changes are included in this PR?

Add more docs and examples for ListArray and OffsetsBuffer

With these changes I think someone has a plausible chance of understanding how to use `ListArray` without having to refer to the Arrow spec

Screenshot 2023-07-31 at 3 15 47 PM
Screenshot 2023-07-31 at 3 15 50 PM

# Are there any user-facing changes?
Better docs

No API changes

### CODE REVIEW

File: arrow-array/src/array/list_array.rs [hunk 1]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///
+/// ```text
+///                                         ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+///                                                                 ┌ ─ ─ ─ ─ ─ ─ ┐    │
+///  ┌─────────────┐  ┌───────┐             │     ┌───┐   ┌───┐       ┌───┐ ┌───┐
+///  │   [A,B,C]   │  │ (0,3) │                   │ 1 │   │ 0 │     │ │ 1 │ │ A │ │ 0  │
+///  ├─────────────┤  ├───────┤             │     ├───┤   ├───┤       ├───┤ ├───┤
+///  │ [] (empty)  │  │ (3,3) │                   │ 1 │   │ 3 │     │ │ 1 │ │ B │ │ 1  │
+///  ├─────────────┤  ├───────┤             │     ├───┤   ├───┤       ├───┤ ├───┤
+///  │    NULL     │  │ (3,4) │                   │ 0 │   │ 4 │     │ │ 1 │ │ C │ │ 2  │
```
Comment (tustvold) [inline arrow-array/src/array/list_array.rs:106]:
```suggestion
///  │    NULL     │  │ (3,4) │                   │ 0 │   │ 3 │     │ │ 1 │ │ C │ │ 2  │
```

File: arrow-array/src/array/list_array.rs [hunk 2]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///  ├─────────────┤  ├───────┤             │     ├───┤   ├───┤       ├───┤ ├───┤
+///  │     [D]     │  │ (4,5) │                   │ 1 │   │ 5 │     │ │ 1 │ │ ? │ │ 3  │
```
Comment (tustvold) [inline arrow-array/src/array/list_array.rs:108]:
```suggestion
///  │     [D]     │  │ (4,5) │                   │ 1 │   │ 4 │     │ 0│  │ │ ? │ │ 3  │
```

File: arrow-array/src/array/list_array.rs [hunk 3]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///  ├─────────────┤  ├───────┤             │     ├───┤   ├───┤       ├───┤ ├───┤
+///  │  [NULL, F]  │  │ (5,7) │                   │ 1 │   │ 5 │     │ │ 0 │ │ D │ │ 4  │
```
Comment (tustvold) [inline arrow-array/src/array/list_array.rs:110]:
```suggestion
///  │  [NULL, F]  │  │ (5,7) │                   │ 1 │   │ 5 │     │ │ 1 │ │ D │ │ 4  │
```

File: arrow-array/src/array/list_array.rs [hunk 4]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///  └─────────────┘  └───────┘             │     └───┘   ├───┤       ├───┤ ├───┤
+///                                                       │ 7 │     │ │ 1 │ │ ? │ │ 5  │
```
Comment (tustvold) [inline arrow-array/src/array/list_array.rs:112]:
```suggestion
///                                                       │ 7 │     │ │ 0 │ │ ? │ │ 5  │
```

File: arrow-array/src/array/list_array.rs [hunk 5]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///                 (offsets[i],            │   ListArray               (Array)
+///                offsets[i+1])                                    └ ─ ─ ─ ─ ─ ─ ┘    │
+///                                         └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+///
+///
+/// ```
+///
+/// # Example of constructing a [`ListArray`]
+///
+/// Here is code that constructs the [`ListArray`' in the example above
```
[the remark this answers is not in the document]
Comment (alamb) [reply arrow-array/src/array/list_array.rs:126]:
```suggestion
/// Here is code that constructs the [`ListArray`] in the example above
```

File: arrow-array/src/array/list_array.rs (hunk 1 above)
Comment (alamb) [reply arrow-array/src/array/list_array.rs:106]: Thank you -- this is why I made the code example.

File: arrow-array/src/array/list_array.rs [hunk 6]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///
+/// // Values is a string array
+/// // "A", "B" "C", "?", "D", NULL, "F"
+/// assert_eq!(
+///   array.values().as_ref(),
+///   &StringArray::from(vec![
+///     Some("A"), Some("B"), Some("C"),
+///     Some("?"), Some("D"), None,
+///     Some("F")
+///   ]) as &dyn Array
```
Comment (tustvold) [inline arrow-array/src/array/list_array.rs:168]:
```suggestion
///   ])
```
It should infer this automatically

**Review by tustvold — APPROVED** [review]

File: arrow-array/src/array/list_array.rs [hunk 7]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+/// [`LargeListArray`]` with `i64` offsets.
+///
+/// Use [`GenericListBuilder`](crate::builder::GenericListBuilder) to
+/// construct a [`GenericListArray`].
+///
+/// # Representation
+///
+/// A [`ListArray`] can represent a list of values of any other
+/// supported Arrow type. Each element of the `ListArray` itself is a
+/// a list which mayb be empty, may contain NULL and non-null values,
```
Comment (viirya) [inline arrow-array/src/array/list_array.rs:74]:
```suggestion
/// a list which maybe empty, may contain NULL and non-null values,
```

File: arrow-array/src/array/list_array.rs [hunk 8]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+/// or may itself be NULL.
+///
+/// For example, this `ListArray` stores lists of strings:
+///
+/// ```text
+/// ┌─────────────┐
+/// │   [A,B,C]   │
```
Comment (viirya) [inline arrow-array/src/array/list_array.rs:81]: To be consistent in style with below:

```suggestion
/// │   [A, B, C]   │
```

File: arrow-buffer/src/buffer/offset.rs [hunk 9]
```diff
@@ -19,7 +19,43 @@ use crate::buffer::ScalarBuffer;
...
+/// [`OffsetBuffer`] are used to represent ranges of offsets. An
+/// `OffsetBuffer` of `N+1` items contains `N` such ranges. The start
+/// offset for element `i` is `offsets[i]` and the end offset is
+/// `offsets[i+1]`. Equal offsets represent an empty range.
+///
+/// # Example
+///
+/// This example shows how 5 distinct ranges, are represented using a
+/// 6 entry `OffsetBuffer`. The first entry `(0, 3)` represents the
+/// three offsets `0, 1, 2` entry `(3,3)` represent no offsets
```
Comment (viirya) [inline arrow-buffer/src/buffer/offset.rs:33]:
```suggestion
/// 6 entry `OffsetBuffer`. The first entry `(0, 3)` represents the
/// three offsets `0, 1, 2`. The entry `(3,3)` represent no offsets
```

File: arrow-buffer/src/buffer/offset.rs [hunk 10]
```diff
@@ -19,7 +19,43 @@ use crate::buffer::ScalarBuffer;
...
+///   │ (3,4) │                │ 4 │
+///   ├───────┤                ├───┤
+///   │ (4,5) │                │ 5 │
+///   ├───────┤                ├───┤
+///   │ (5,7) │                │ 5 │
+///   └───────┘                ├───┤
+///                            │ 7 │
+///                            └───┘
+///
+///                        Offsets Buffer
```
Comment (viirya) [inline arrow-buffer/src/buffer/offset.rs:51]:
```suggestion
///   ┌───────┐                ┌───┐
///   │ (0,3) │                │ 0 │
///   ├───────┤                ├───┤
///   │ (3,3) │                │ 3 │
///   ├───────┤                ├───┤
///   │ (3,4) │                │ 3 │
///   ├───────┤                ├───┤
///   │ (4,5) │                │ 4 │
///   ├───────┤                ├───┤
///   │ (5,7) │                │ 5 │
///   └───────┘                ├───┤
///                            │ 7 │
///                            └───┘
///
///                        Offsets Buffer
```

**Review by viirya — APPROVED** [review]

File: arrow-buffer/src/buffer/offset.rs (hunk 10 above)
Comment (alamb) [reply arrow-buffer/src/buffer/offset.rs:51]: Thank you

**Comment by alamb** [comment]
I'll merge this later in the week if there are no more comments

cc @jayzhan211 @smiklos  and @vincev  who may also be interested

**Comment by smiklos** [comment]
Would it make sense to also document `FixedSizeListArray` ? It seems to store values differently than `GenericListArray`

File: arrow-array/src/array/list_array.rs [hunk 11]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+/// a list which may be empty, may contain NULL and non-null values,
+/// or may itself be NULL.
+///
+/// For example, this `ListArray` stores lists of strings:
+///
+/// ```text
+/// ┌─────────────┐
+/// │   [A,B,C]   │
+/// ├─────────────┤
+/// │ [] (empty)  │
```
Comment (jayzhan211) [inline arrow-array/src/array/list_array.rs:83]: This is a little confusing to me.
`(empty)` is just the explanation to `[]`?

**Comment by smiklos** [comment]
> > FixedSizeListArray stores value the same as ListArray but each row has the same size of array.
> 
> This isn't entirely accurate, there is no set of offsets which means that all slots have the same width. In particular this includes nulls, which can cause some interesting challenges.

Based on the following array 
```
[[1, 2], Null, [3,4] 
```

GenericListArray::values will look like this 

```
[1, 2, 3]
```

But FixedSizeArray::values would look like this

```
[1, 2, Null, Null, 3, 4]

File: arrow-array/src/array/list_array.rs (hunk 11 above)
Comment (jayzhan211) [reply arrow-array/src/array/list_array.rs:83]: I think just leaving `[]` might be better.

**Comment by alamb** [comment]
> Would it make sense to also document FixedSizeListArray ? It seems to store values differently than GenericListArray

Yes it absolutely would make sense. If you have time that would be great. Otherwise I'll try and make a diagram tomorrow

File: arrow-array/src/array/list_array.rs (hunk 11 above)
Comment (alamb) [reply arrow-array/src/array/list_array.rs:83]: Good idea -- I will make that change

File: arrow-array/src/array/list_array.rs [hunk 12]
```diff
@@ -54,11 +54,129 @@ impl OffsetSizeTrait for i64 {
...
+///
+/// ```
+/// # use std::sync::Arc;
+/// # use arrow_array::{builder::ListBuilder, builder::StringBuilder, ArrayRef, StringArray, Array};
+/// #
+/// let values_builder = StringBuilder::new();
+/// let mut builder = ListBuilder::new(values_builder);
+///
+/// // [A, B, C]
+/// builder.values().append_value("A");
```
[the remark this answers is not in the document]
Comment (alamb) [reply arrow-array/src/array/list_array.rs]: I was torn on this, as keeping the example (that constructs the data in the diagram) I thought added value. However I think your idea of adding a link to the approach would be good. 

I implemented this change in 151564d5ab

**Pull request merged**